File: backend/app/schemas/exam.py
from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel


class ExamPdfRequest(BaseModel):
    question_ids: list[int]
    title: str = "수학 시험지"
    subtitle: Optional[str] = None
    include_source: bool = True
    include_logo: bool = False
    overrides: dict[int, str] = {}
    preserve_order: bool = False


class BookPdfRequest(BaseModel):
    question_ids: list[int]
    title: str = "수학 교재"
    subtitle: Optional[str] = None
    include_source: bool = True
    include_logo: bool = False
    cover_style: Literal["final", "diagonal"] = "final"
    cover_kicker: Optional[str] = None
    cover_big_word: Optional[str] = None
    cover_footer_main: Optional[str] = None
    cover_footer_sub: Optional[str] = None
    dcov_subject: Optional[str] = None
    dcov_level: Optional[str] = None
    kicker_mark: Optional[str] = None
    kicker_text: Optional[str] = None
    divider_meta_top: Optional[str] = None
    divider_footer_title: Optional[str] = None
    divider_footer_sub: Optional[str] = None
    overrides: dict[int, str] = {}
    book_mode: Literal["chapter", "flat"] = "chapter"
    flat_layout: Literal["half", "full"] = "half"
    preserve_order: bool = False


class LayoutPreviewRequest(BaseModel):
    question_ids: list[int]
    overrides: dict[int, str] = {}
    preserve_order: bool = False


class LayoutSlot(BaseModel):
    question_id: int
    layout: Literal["half", "full"]


class LayoutColumn(BaseModel):
    slots: list[LayoutSlot]


class LayoutPage(BaseModel):
    columns: list[LayoutColumn]


class LayoutPreviewResponse(BaseModel):
    pages: list[LayoutPage]

File: frontend/src/api/exam.ts
import { apiFetchBlob, apiFetchText } from "./client";

export interface ExamPdfRequest {
  question_ids: number[];
  title: string;
  subtitle?: string | null;
  include_source: boolean;
  include_logo: boolean;
  preserve_order?: boolean;
}

export function downloadExamPdf(req: ExamPdfRequest): Promise<Blob> {
  return apiFetchBlob("/api/exam/pdf", {
    method: "POST",
    body: JSON.stringify(req),
  });
}

// 실물 미리보기용 — Playwright 없이 실제 PDF와 같은 HTML을 그대로 받아
// iframe에 넣는다(다운로드 집계에 안 잡히도록 별도 엔드포인트).
export function fetchExamHtmlPreview(req: ExamPdfRequest): Promise<string> {
  return apiFetchText("/api/exam/html-preview", {
    method: "POST",
    body: JSON.stringify(req),
  });
}

export type CoverStyle = "final" | "diagonal";
export type BookMode = "chapter" | "flat";
export type FlatLayout = "half" | "full";

export interface BookPdfRequest {
  question_ids: number[];
  title: string;
  subtitle?: string | null;
  include_source: boolean;
  include_logo: boolean;
  cover_style: CoverStyle;
  cover_kicker?: string | null;
  cover_big_word?: string | null;
  cover_footer_main?: string | null;
  cover_footer_sub?: string | null;
  dcov_subject?: string | null;
  dcov_level?: string | null;
  kicker_mark?: string | null;
  kicker_text?: string | null;
  divider_meta_top?: string | null;
  divider_footer_title?: string | null;
  divider_footer_sub?: string | null;
  book_mode?: BookMode;
  flat_layout?: FlatLayout;
}

export function downloadBookPdf(req: BookPdfRequest): Promise<Blob> {
  return apiFetchBlob("/api/exam/book-pdf", {
    method: "POST",
    body: JSON.stringify(req),
  });
}

export function fetchBookHtmlPreview(req: BookPdfRequest): Promise<string> {
  return apiFetchText("/api/exam/book-html-preview", {
    method: "POST",
    body: JSON.stringify(req),
  });
}

